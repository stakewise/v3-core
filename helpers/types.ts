import { BigNumberish } from 'ethers'
export type ThenArg<T> = T extends PromiseLike<infer U> ? U : T

export enum Networks {
  mainnet = 'mainnet',
  goerli = 'goerli',
}

export type NetworkConfig = {
  url: string
  chainId: number

  governor: string
  validatorsRegistry: string
  securityDeposit: BigNumberish

  // Keeper
  oracles: string[]
  rewardsMinOracles: number
  validatorsMinOracles: number
  rewardsDelay: number
  oraclesConfigIpfsHash: string
  maxAvgRewardPerSecond: BigNumberish

  // OsToken
  treasury: string
  osTokenFeePercent: number
  osTokenCapacity: BigNumberish
  osTokenName: string
  osTokenSymbol: string

  // OsTokenConfig
  redeemFromLtvPercent: number
  redeemToLtvPercent: number
  liqThresholdPercent: number
  liqBonusPercent: number
  ltvPercent: number

  // GenesisEthVault
  genesisVault: {
    admin: string
    poolEscrow: string
    stakedEthToken: string
    capacity: BigNumberish
    feePercent: number
  }
}
