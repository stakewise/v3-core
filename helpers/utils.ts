import { Contract } from 'ethers'
import { HardhatRuntimeEnvironment } from 'hardhat/src/types/runtime'

export async function deployContract(tx: any): Promise<Contract> {
  const result = await tx
  await result.deployTransaction.wait()
  return result
}

async function delay(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

export async function verify(
  hre: HardhatRuntimeEnvironment,
  address: string,
  constructorArgs: any,
  contractPath: string
) {
  if (!process.env.BLOCK_EXPLORER_KEY) {
    return
  }

  let count = 0
  const maxTries = 8
  // eslint-disable-next-line no-constant-condition
  while (true) {
    await delay(10000)
    try {
      console.log('Verifying contract at', address)
      await hre.run('verify:verify', {
        address,
        constructorArguments: constructorArgs,
        contract: contractPath,
      })
      break
    } catch (error) {
      if (String(error).includes('Already Verified')) {
        console.log(`Already verified contract at ${contractPath} at address ${address}`)
        break
      }
      if (++count == maxTries) {
        console.log(
          `Failed to verify contract at ${contractPath} at address ${address}, error: ${error}`
        )
        break
      }
      console.log(`Retrying... Retry #${count}, last error: ${error}`)
    }
  }
}
