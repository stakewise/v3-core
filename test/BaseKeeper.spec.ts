import { ethers, waffle } from 'hardhat'
import { BigNumber, Wallet } from 'ethers'
import { hexlify, parseEther } from 'ethers/lib/utils'
import { EthVault, EthKeeper, Oracles } from '../typechain-types'
import { ThenArg } from '../helpers/types'
import { ethVaultFixture } from './shared/fixtures'
import { expect } from './shared/expect'
import { ORACLES, ZERO_ADDRESS, ZERO_BYTES32 } from './shared/constants'
import snapshotGasCost from './shared/snapshotGasCost'
import {
  createVaultRewardsRoot,
  getRewardsRootProof,
  RewardsRoot,
  VaultReward,
} from './shared/rewards'
import { setBalance } from './shared/utils'

const createFixtureLoader = waffle.createFixtureLoader

describe('BaseKeeper', () => {
  const capacity = parseEther('1000')
  const feePercent = 1000
  const name = 'SW ETH Vault'
  const symbol = 'SW-ETH-1'
  const validatorsRoot = '0x059a8487a1ce461e9670c4646ef85164ae8791613866d28c972fb351dc45c606'
  const validatorsIpfsHash = '/ipfs/QmfPnyNojfyqoi9yqS3jMp16GGiTQee4bdCXJC64KqvTgc'
  const metadataIpfsHash = '/ipfs/QmanU2bk9VsJuxhBmvfgXaC44fXpcC8DNHNxPZKMpNXo37'

  let loadFixture: ReturnType<typeof createFixtureLoader>
  let createVault: ThenArg<ReturnType<typeof ethVaultFixture>>['createVault']
  let getSignatures: ThenArg<ReturnType<typeof ethVaultFixture>>['getSignatures']

  let sender: Wallet, owner: Wallet, admin: Wallet, oracle: Wallet
  let keeper: EthKeeper, oracles: Oracles, vault: EthVault

  before('create fixture loader', async () => {
    ;[sender, admin, owner] = await (ethers as any).getSigners()
    loadFixture = createFixtureLoader([owner])
    oracle = new Wallet(ORACLES[0], await waffle.provider)
  })

  beforeEach(async () => {
    ;({ oracles, keeper, createVault, getSignatures } = await loadFixture(ethVaultFixture))
    vault = await createVault(admin, {
      capacity,
      validatorsRoot,
      feePercent,
      name,
      symbol,
      validatorsIpfsHash,
      metadataIpfsHash,
    })
    await setBalance(oracle.address, parseEther('10000'))
  })

  describe('set rewards root', () => {
    let vaultReward: VaultReward
    let rewardsRoot: RewardsRoot

    beforeEach(async () => {
      vaultReward = { reward: parseEther('5'), vault: vault.address }
      rewardsRoot = createVaultRewardsRoot([vaultReward], oracles)
    })

    it('fails with invalid root', async () => {
      await expect(
        keeper
          .connect(oracle)
          .setRewardsRoot(
            ZERO_BYTES32,
            rewardsRoot.updateTimestamp,
            rewardsRoot.ipfsHash,
            getSignatures(rewardsRoot.signingData)
          )
      ).to.be.revertedWith('InvalidRewardsRoot()')
    })

    it('fails if not oracle', async () => {
      await expect(
        keeper
          .connect(sender)
          .setRewardsRoot(
            rewardsRoot.root,
            rewardsRoot.updateTimestamp,
            rewardsRoot.ipfsHash,
            getSignatures(rewardsRoot.signingData)
          )
      ).to.be.revertedWith('AccessDenied()')
    })

    it('fails with invalid IPFS hash', async () => {
      await expect(
        keeper
          .connect(oracle)
          .setRewardsRoot(
            rewardsRoot.root,
            rewardsRoot.updateTimestamp,
            ZERO_BYTES32,
            getSignatures(rewardsRoot.signingData)
          )
      ).to.be.revertedWith('InvalidOracle()')
    })

    it('fails with invalid nonce', async () => {
      const signatures = getSignatures(rewardsRoot.signingData)
      await keeper
        .connect(oracle)
        .setRewardsRoot(
          rewardsRoot.root,
          rewardsRoot.updateTimestamp,
          rewardsRoot.ipfsHash,
          signatures
        )

      const newVaultReward = { reward: parseEther('3'), vault: vault.address }
      const newRewardsRoot = createVaultRewardsRoot([newVaultReward], oracles)
      await expect(
        keeper
          .connect(oracle)
          .setRewardsRoot(
            newRewardsRoot.root,
            rewardsRoot.updateTimestamp,
            newRewardsRoot.ipfsHash,
            getSignatures(newRewardsRoot.signingData)
          )
      ).to.be.revertedWith('InvalidOracle()')
    })

    it('succeeds', async () => {
      const signatures = getSignatures(rewardsRoot.signingData)
      const receipt = await keeper
        .connect(oracle)
        .setRewardsRoot(
          rewardsRoot.root,
          rewardsRoot.updateTimestamp,
          rewardsRoot.ipfsHash,
          signatures
        )

      await expect(receipt)
        .to.emit(keeper, 'RewardsRootUpdated')
        .withArgs(
          oracle.address,
          rewardsRoot.root,
          rewardsRoot.updateTimestamp,
          1,
          rewardsRoot.ipfsHash,
          hexlify(signatures)
        )
      expect(await keeper.rewardsRoot()).to.eq(rewardsRoot.root)
      expect(await keeper.rewardsNonce()).to.eq(2)
      await snapshotGasCost(receipt)
    })
  })

  describe('is harvested', () => {
    let vaultReward: VaultReward
    let rewardsRoot: RewardsRoot
    let proof: string[]

    beforeEach(async () => {
      vaultReward = { reward: parseEther('5'), vault: vault.address }
      rewardsRoot = createVaultRewardsRoot([vaultReward], oracles)
      const signatures = getSignatures(rewardsRoot.signingData)
      await keeper
        .connect(oracle)
        .setRewardsRoot(
          rewardsRoot.root,
          rewardsRoot.updateTimestamp,
          rewardsRoot.ipfsHash,
          signatures
        )
      proof = getRewardsRootProof(rewardsRoot.tree, vaultReward)
    })

    it('returns true for uncollateralized vault', async () => {
      expect(await keeper.isHarvested(vault.address)).to.equal(true)
    })

    it('returns false for collateralized unharvested vault', async () => {
      await keeper.harvest(vaultReward.vault, vaultReward.reward, proof)
      let newVaultReward = { reward: parseEther('3'), vault: vault.address }
      let newRewardsRoot = createVaultRewardsRoot(
        [newVaultReward],
        oracles,
        rewardsRoot.updateTimestamp,
        2
      )
      await keeper
        .connect(oracle)
        .setRewardsRoot(
          newRewardsRoot.root,
          rewardsRoot.updateTimestamp,
          newRewardsRoot.ipfsHash,
          getSignatures(newRewardsRoot.signingData)
        )

      // returns true if not harvested one time
      expect(await keeper.isHarvested(vault.address)).to.equal(true)

      const newTimestamp = rewardsRoot.updateTimestamp + 1
      newVaultReward = { reward: parseEther('4'), vault: vault.address }
      newRewardsRoot = createVaultRewardsRoot([newVaultReward], oracles, newTimestamp, 3)
      await keeper
        .connect(oracle)
        .setRewardsRoot(
          newRewardsRoot.root,
          newTimestamp,
          newRewardsRoot.ipfsHash,
          getSignatures(newRewardsRoot.signingData)
        )

      // returns false if not harvested two times
      expect(await keeper.isHarvested(vault.address)).to.equal(false)
    })

    it('returns true for collateralized harvested vault', async () => {
      await keeper.harvest(vaultReward.vault, vaultReward.reward, proof)
      const newVaultReward = { reward: parseEther('3'), vault: vault.address }
      const newRewardsRoot = createVaultRewardsRoot(
        [newVaultReward],
        oracles,
        rewardsRoot.updateTimestamp,
        2
      )
      await keeper
        .connect(oracle)
        .setRewardsRoot(
          newRewardsRoot.root,
          rewardsRoot.updateTimestamp,
          newRewardsRoot.ipfsHash,
          getSignatures(newRewardsRoot.signingData)
        )
      await keeper.harvest(
        newVaultReward.vault,
        newVaultReward.reward,
        getRewardsRootProof(newRewardsRoot.tree, newVaultReward)
      )
      expect(await keeper.isHarvested(vault.address)).to.equal(true)
    })
  })

  describe('harvest', () => {
    let vaultReward: VaultReward
    let rewardsRoot: RewardsRoot
    let proof: string[]

    beforeEach(async () => {
      vaultReward = { reward: parseEther('5'), vault: vault.address }

      const vaultRewards = [vaultReward]
      for (let i = 1; i < 11; i++) {
        const vlt = await createVault(admin, {
          capacity,
          validatorsRoot,
          feePercent,
          name,
          symbol,
          validatorsIpfsHash,
          metadataIpfsHash,
        })
        vaultRewards.push({ reward: parseEther(i.toString()), vault: vlt.address })
      }

      rewardsRoot = createVaultRewardsRoot(vaultRewards, oracles)
      proof = getRewardsRootProof(rewardsRoot.tree, vaultReward)
      await keeper
        .connect(oracle)
        .setRewardsRoot(
          rewardsRoot.root,
          rewardsRoot.updateTimestamp,
          rewardsRoot.ipfsHash,
          getSignatures(rewardsRoot.signingData)
        )
    })

    it('fails for invalid vault', async () => {
      await expect(keeper.harvest(ZERO_ADDRESS, vaultReward.reward, proof)).to.be.revertedWith(
        'InvalidVault()'
      )
    })

    it('fails for invalid reward', async () => {
      await expect(keeper.harvest(vaultReward.vault, 0, proof)).to.be.revertedWith('InvalidProof()')
    })

    it('fails for invalid proof', async () => {
      await expect(keeper.harvest(vaultReward.vault, vaultReward.reward, [])).to.be.revertedWith(
        'InvalidProof()'
      )
    })

    it('calculates delta since last update', async () => {
      await keeper.harvest(vaultReward.vault, vaultReward.reward, proof)

      const newVaultReward = {
        reward: BigNumber.from(vaultReward.reward).sub(parseEther('1')),
        vault: vaultReward.vault,
      }

      const newRewardsRoot = createVaultRewardsRoot(
        [newVaultReward],
        oracles,
        rewardsRoot.updateTimestamp,
        2
      )
      const signatures = getSignatures(newRewardsRoot.signingData)
      await keeper
        .connect(oracle)
        .setRewardsRoot(
          newRewardsRoot.root,
          rewardsRoot.updateTimestamp,
          newRewardsRoot.ipfsHash,
          signatures
        )

      const newProof = getRewardsRootProof(newRewardsRoot.tree, newVaultReward)
      const receipt = await keeper
        .connect(sender)
        .harvest(newVaultReward.vault, newVaultReward.reward, newProof)
      await expect(receipt)
        .to.emit(keeper, 'Harvested')
        .withArgs(sender.address, newVaultReward.vault, newVaultReward.reward)
      await expect(receipt)
        .to.emit(vault, 'StateUpdated')
        .withArgs(BigNumber.from(newVaultReward.reward).sub(vaultReward.reward))
      await snapshotGasCost(receipt)
    })

    it('succeeds for already harvested vault', async () => {
      await keeper.harvest(vaultReward.vault, vaultReward.reward, proof)
      const receipt = await keeper
        .connect(sender)
        .harvest(vaultReward.vault, vaultReward.reward, proof)
      await expect(receipt).to.not.emit(keeper, 'Harvested')
      await expect(receipt).to.emit(vault, 'StateUpdated').withArgs(0)
      await snapshotGasCost(receipt)
    })

    it('succeeds for not harvested vault', async () => {
      const receipt = keeper.connect(sender).harvest(vaultReward.vault, vaultReward.reward, proof)
      await expect(receipt)
        .to.emit(keeper, 'Harvested')
        .withArgs(sender.address, vaultReward.vault, vaultReward.reward)
      await expect(receipt).to.emit(vault, 'StateUpdated').withArgs(vaultReward.reward)
      await snapshotGasCost(receipt)
    })
  })

  describe('upgrade', () => {
    it('fails for not an owner', async () => {
      await expect(keeper.connect(sender).upgradeTo(keeper.address)).to.revertedWith(
        'Ownable: caller is not the owner'
      )
    })
    it('succeeds for owner', async () => {
      await expect(keeper.connect(owner).upgradeTo(keeper.address)).to.revertedWith(
        'ERC1967Upgrade: new implementation is not UUPS'
      )
    })
  })
})
