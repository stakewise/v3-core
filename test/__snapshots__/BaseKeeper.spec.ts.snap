// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseKeeper harvest calculates delta since last update 1`] = `
Object {
  "calldataByteLength": 132,
  "gasUsed": 67912,
}
`;

exports[`BaseKeeper harvest succeeds for already harvested vault 1`] = `
Object {
  "calldataByteLength": 228,
  "gasUsed": 60396,
}
`;

exports[`BaseKeeper harvest succeeds for not harvested vault 1`] = `
Object {
  "calldataByteLength": 228,
  "gasUsed": 130849,
}
`;

exports[`BaseKeeper set rewards root succeeds 1`] = `
Object {
  "calldataByteLength": 676,
  "gasUsed": 132370,
}
`;
